melodies_monet/util/region_select.py

```diff
@@ -0,0 +1,320 @@
+# SPDX-License-Identifier: Apache-2.0
+"""
+Masking arbitrary regions with regionmask
+"""
+
+import re
+import warnings
+
+import pandas as pd
+import requests
+
+from melodies_monet.util.tools import get_epa_region_bounds, get_giorgi_region_bounds
+
+try:
+    import geopandas as gp
+    from shapely.geometry import MultiPolygon, Polygon
+except ImportError:
+    gp = None
+try:
+    import regionmask
+except ImportError:
+    regionmask = None
+
+
+def _download_with_name(url, verbose=False):
+    """Automates download while reading content disposition if needed.
+
+    Parameters:
+    -----------
+    url : str
+        URL of file to download
+    verbose : boolean
+        Whether to add verbosity
+    Returns
+    -------
+    str
+       Path to downloaded file
+    """
+    r = requests.get(url, allow_redirects=True, timeout=10)
+    fname = re.findall("filename=(.+)", r.headers.get("content-disposition"))[0]
+    if len(fname) == 0:
+        fname = url.rsplit("/", 1)[1]
+    if fname[0] == '"':
+        fname = fname[1:-1]
```

Comment:
As it was, if Content-Disposition was missing the `re.findall` will error. And if it was there but `filename=` wasn't in it, `[0]` would error. And if `filename` wasn't last, the match would've included other directives. This should be a bit more robust:
```suggestion
    fname = None
    content_disposition = r.headers.get("Content-Disposition")
    if content_disposition:
        m = re.search(r"filename=(.+)[;$]", content_disposition)
        if m is not None:
            fname = m.group(1).strip('"')
    if fname is None:
        fname = url.rsplit("/", 1)[-1]
```

Reply:
Would `filename=([^;]+)(?=;|$)"` work better for the filename regex? [`$` in a character class will match only a $ in the input](https://docs.python.org/3/howto/regex.html#more-metacharacters) while this looks like it's trying to match end-of-line, and [`(?=` starts a lookahead assertion](https://docs.python.org/3/howto/regex.html#lookahead-assertions) (as an alternative, [`(?:` starts a non-capturing group](https://docs.python.org/3/howto/regex.html#non-capturing-and-named-groups) that won't appear in `m.groups()`)  I don't see anything there to keep the `.+` from gobbling the next item: [the non-greedy `.+?` should also work](https://docs.python.org/3/howto/regex.html#greedy-versus-non-greedy)

Alternately, `[part.split("=")[1] for part in content_disposition.split(";") if "filename=" in part]`

[This page](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition) suggests the server might provide `filname*` in addition to or instead of `filename`, though that value would require additional processing.

It doesn't look like pooch has a way to set the file name after the download has started, but you might be able to guess from the URL.   Unfortunately, there's enough variation in how the URL encodes information about the resource it points to that a universal method of composing a simple filename obviously related to the URL is probably not possible.

Pooch's default involves a hash of the URL, would using just the hash without the guess at the filename work?  It's not as obviously related to the URL if you want to find the downloaded file in pooch's cache, but the code will be able to reconstruct it easily.

Reply:
Thanks @DWesl for the help and explanations. Turns out I hadn't properly checked the end-of-line case. Updated to the non-greedy with (non-capturing) group for the end.

Reply:
I'm not very good with regex. I just committed the last version from @zmoon, does it need to match @DWesl ? I can update it if required.

Reply:
@blychs the one you committed should be fine, that was based on @DWesl comments.

Reply:
Going to resolve this and move the pooch discussion to main thread.

Reply:
> It doesn't look like pooch has a way to set the file name after the download has started, but you might be able to guess from the URL. Unfortunately, there's enough variation in how the URL encodes information about the resource it points to that a universal method of composing a simple filename obviously related to the URL is probably not possible.

This does not always work. E. g., if you look at the example in the docs that I added, the state of Colorado has URL's for their shapefiles that don't show the name at all. It is more than good enough in _most_ cases, but that few that don't work are, IMHO, bad enough that it's not the best for this generalized tool.


> Pooch's default involves a hash of the URL, would using just the hash without the guess at the filename work? It's not as obviously related to the URL if you want to find the downloaded file in pooch's cache, but the code will be able to reconstruct it easily. It might be worth adding some part of the domain/hostname in the hash to get a little association between URL and cached file name and to reduce the chance of filename collisions.

I am not sure if that would work, though. Regionmask assumes the format from the extension if I'm not wrong, and I'm definitely doing that for zipped files. That being the case, if I don't have a name containing the extension, this would fail, I believe?